enhancements/monitoring/restart-metrics.md

```diff
@@ -0,0 +1,105 @@
+---
+title: restart-metrics
+authors:
+  - "@rphillips"
+reviewers:
+  - "@smarterclayton"
+  - "@sjenning"
+  - "@mrunalp"
+approvers:
+  - TBD
+creation-date: 2020-03-23
+last-updated: 2020-03-23
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Service Restart Metrics
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+There is not any insight with systemd units restarting. Metrics of this sort are
+important to know if Kubelet, crio, or other system services are crashing and
+restarting. Restart metrics are going to be extremely important since the
+Kubelet is going to be changed to exit on a crash, thus restarting. Previous
+behavior of the Kubelet is to recover() from the crash and not exit.
+
+## Motivation
+
+Systemd unit restart metrics of a unit are vitally important for system
+administrators to see the health of the Kubelet, crio, and other vitally important
+services.
+
+### Goals
+
+- Systemd unit restart metrics need to be propogated through to monitoring, and
+  the alerting system to alert system administrators to problems within the
+  system.
+
+### Non-Goals
+
+### Current Behavior
+
+node_exporter is running in a non-privileged container - as user `nobody` - and
+cannot connect to the DBUS socket to communicate with systemd. Monitoring has
+concerns the underlying [systemd
+collector](https://github.com/prometheus/node_exporter/blob/master/collector/systemd_linux.go)
+is not performant enough.
+
+## Proposal
+
+- This proposal is to raise the privileges on node_exporter.
+
+- Whitelist in the systemd node_exporter collector: crio, kubelet, and any other
+  agreed to system service.
+
+### Risks and Mitigations
+
+We could have a sidecar container running privileged that injects metrics into
+node_exporter; however, increasing privileges on node_exporter is a valid
```

Comment:
> however, increasing privileges on node_exporter is a valid
option. A new sidecar container would add another dependency and service that
would need to be installed and maintained.

While I appreciate pointing out the need for maintaining the sidecar, I would prefer to rather restrict security privileges vs maintaince.

Reply:
I do not see the difference of running a sidecar as privileged, rather than running the main component as privileged ?

Reply:
I worry about adding more processes to the system, which take up more resources (cpu and memory limits). Granted that systemd_exporter is likely lightweight, but it is one added process.

Reply:
A new sidecar doesn't require a dependency, it can be the same binary but restricted.  The goal of the sidecar is to firewall high level permissions and prevent accidental escape.

Note though that all code that runs privileged has to be reviewed at roughly the same level.  Node exporter is a bigger surface area than a smaller subset of code reading a smaller subset of metrics.

Reply:
OK. I'll add into the proposal to run systemd_exporter as a sidecar privileged container.